### PR TITLE
Fix #7765, Fix #7618: Portfolio Group By filter & Empty State

### LINKED ISSUE #7618: Portfolio Filters Empty State(s)
### Description: 

Title: `No available assets`
Description: `Deposit or purchase tokens to get started. If you don't see tokens from an imported account, check the filters and display settings. Unknown tokens may need to be added as custom assets.`

https://bravesoftware.slack.com/archives/C023VS4HJ6Q/p1686939237810999

### LINKED ISSUE #7765: Support Portfolio Group By Filter
### Description: 

Support `Group by` filter in Portfolio. 
Title:  `Group by`
Description: `Group assets by`
Options: `None` (default), `Accounts`, `Networks`

If a group has no tokens visible (after other filters are applied), it should be hidden. Ex. Hiding small balances & grouping by account, an account group with no balance for any token would be hidden.

Portfolio v2 designs have grouping support, if done with Portfolio v1 use section grouping.

### PR DESCRIPTION
**Pull request by StephenHeaps**
## Summary of Changes
- Implements the `Group By` filter with `none` (default), `accounts` and `networks` grouping options.
- Implements loading state when loading balances. 
  - Loading state is only shown if hiding small balances, OR assets are grouped (groups with 0 total balance are hidden)

This pull request fixes #7765, fixes #7618

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`
- [x] New or updated UI has been tested across:
  - [x] Light & dark mode
  - [x] Different size classes (iPhone, landscape, iPad)
  - [x] Different dynamic type sizes

## Test Plan:
1. Open Wallet Portfolio, it should not be grouped by default.
2. Tap Filters & Display Settings button and then tap `Reset`
3. Verify default `Group By` option is `None`.
4. Change `Group By` to accounts and tap `Save Changes`.
5. Verify Portfolio is now grouped by each account, with the assets below showing the balance only for that account.
    - Note: Groups with 0 total balance are hidden automatically if `Hide Small Balances` is enabled.
6. Verify groups are sorted by fiat value. 
7. Tap Filters & Display Settings button, then de-select some accounts and tap `Save Changes`
8. Verify de-selected account groups are no longer shown.
9. Tap Filters & Display Settings button then tap `Reset` (re-selects all accounts again), then change `Group By` to `Networks` and tap `Save Changes`
10. Verify Portfolio is now grouped by each network with only assets for that network shown in each group.
    - Note: Groups with 0 total balance are hidden automatically if `Hide Small Balances` is enabled.
11. Verify groups are sorted by fiat value. 
12. Tap Filters & Display Settings button then tap `Reset`.
13. De-select all accounts, or all networks and tap `Save Changes`.
14. Verify empty state is shown.
    - Empty state should be shown whenever no assets are displayed. This can occur with many combinations, for example hiding small balances, grouping assets and then hiding accounts with balances or hiding networks with balances, etc. 
15. Close the wallet, or quit Brave. Re-open Portfolio and verify `Group By` filter is preserved.

## Screenshots:

https://github.com/brave/brave-ios/assets/5314553/6d6e5536-53c2-4a78-8ce4-cf0e93a099b8

## Reviewer Checklist:

- [ ] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `bug` / `enhancement`
- [ ] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [ ] Adequate unit test coverage exists to prevent regressions.
- [ ] Adequate test plan exists for QA to validate (if applicable).
- [ ] Issue and pull request is assigned to a milestone (should happen at merge time).

### CODE REVIEW

**Comment by nuo-xu** [comment]
is it possible to substitute the disclosure symbol?
simulator_screenshot_4213928C-DFCB-4C1A-AC45-6548149A6E72 | Simulator Screenshot - iPhone 14 - 2023-08-04 at 09 46 17 | Screenshot 2023-08-04 at 9 47 30 AM
--|--|--
*Current*|*Similar in Wallet*|*Desktop*

**Comment by nuo-xu** [comment]
once assets are grouped, some rows have a divider but some don't (in the screenshot, there is no divider under `Wrapped Ether`) maybe related to the padding of each row.
Screenshot 2023-08-04 at 10 02 11 AM

**Comment by nuo-xu** [comment]
want to mention here again. 

> Note: Groups with 0 total balance are hidden automatically.

looks like a bug. wdyt? cc @jamesmudgett @srirambv

**Comment by nuo-xu** [comment]
not sure if this is what we want. but its a minor issue. 
for assets with zero balance will be re-ordered (core-data doesn't gives same order), if i tap save changes with no actual changes. 
https://github.com/brave/brave-ios/assets/1187676/d52fa84a-026e-4535-9799-734f0b3ffc86

File: Sources/BraveWallet/Crypto/Portfolio/PortfolioView.swift [hunk 1]
```diff
@@ -213,6 +176,174 @@ struct PortfolioView: View {
...
+            SkeletonLoadingAssetView()
+          } else { // isShowingAssetsEmptyState
+            emptyAssetsState
+          }
+        }
+        .listRowBackground(Color(.secondaryBraveGroupedBackground))
+      }, header: {
+        assetSectionsHeader
+      })
+    } else {
```
Comment (nuo-xu) [inline Sources/BraveWallet/Crypto/Portfolio/PortfolioView.swift:228]: can't we do if/else if?

File: Sources/BraveWallet/Crypto/Portfolio/PortfolioView.swift [hunk 2]
```diff
@@ -213,6 +176,174 @@ struct PortfolioView: View {
...
+  /// Builds the in-section header for an AssetGroupViewModel that is shown in expanded and non-expanded state. Not used for ungrouped assets.
+  private func groupHeader(for group: AssetGroupViewModel) -> some View {
+    VStack(spacing: 0) {
+      HStack {
+        if case let .network(networkInfo) = group.groupType {
+          NetworkIcon(network: networkInfo, length: 32)
+        } else if case let .account(accountInfo) = group.groupType {
+          Blockie(address: accountInfo.address, shape: .rectangle)
+            .frame(width: 32, height: 32)
+            .clipShape(RoundedRectangle(cornerRadius: 4))
```
Comment (nuo-xu) [inline Sources/BraveWallet/Crypto/Portfolio/PortfolioView.swift:324]: i think we should align the account blockie shape through out the wallet.

Comment (StephenHeaps) [reply Sources/BraveWallet/Crypto/Portfolio/PortfolioView.swift:324]: The new v2 designs for other views might have square blockies as well but I don't think we should be updating the entire wallet in this Portfolio Grouping PR.

File: Sources/BraveWallet/Crypto/Portfolio/PortfolioView.swift (hunk 1 above)
Comment (StephenHeaps) [reply Sources/BraveWallet/Crypto/Portfolio/PortfolioView.swift:228]: I'm not sure what you mean here? Adding `else if` here instead would just be unnecessary code:
```
if portfolioStore.isShowingAssetsLoadingState || portfolioStore.isShowingAssetsEmptyState {
...
} else if !portfolioStore.isShowingAssetsLoadingState && !portfolioStore.isShowingAssetsEmptyState {
...
}
```

**Comment by StephenHeaps** [comment]
> is it possible to substitute the disclosure symbol?

Add a custom `WalletDisclosureGroup` for this as `DisclosureGroupStyle` is iOS 16+. [a94a5dd](https://github.com/brave/brave-ios/pull/7826/commits/a94a5dddaec6a0103ba7635692fc26b10b75bddc)

> once assets are grouped, some rows have a divider but some don't (in the screenshot, there is no divider under Wrapped Ether) maybe related to the padding of each row.

Padding and separator issue should be resolved now.

> want to mention here again.
> Note: Groups with 0 total balance are hidden automatically.
> looks like a bug. wdyt?

Groups are now only hidden here now when `Hide Small Balances` is enabled. [eac4c70](https://github.com/brave/brave-ios/pull/7826/commits/eac4c7038e5d0bb338b89a4a542fb2c06c384775)

> not sure if this is what we want. but its a minor issue.
for assets with zero balance will be re-ordered (core-data doesn't gives same order), if i tap save changes with no actual changes.

Updated sort logic when sorting empty balances by value to move native token(s) to top of list, otherwise will sort alphabetically when 2+ assets have 0 balance [92eadf2](https://github.com/brave/brave-ios/pull/7826/commits/92eadf2551388de62df4951a86fa51fc40e782c9).

**Comment by nuo-xu** [comment]
all my comments have been addressed. Portfolio looks pretty awesome now! good work here!

**Review by nuo-xu — APPROVED** [review]

**Pull request merged**